### PR TITLE
Camera Tracking

### PR DESCRIPTION
**Pull request by Declipsonator**
# Camera Tracking
## AprilTags.java
See all the AprilTags around the robot. Supports multiple cameras (takes the one with the highest ambiguity) and has a few functions to find the closest target/the target with the best ambiguity.
## RobotFieldPosition.java
Small class to see where the robot is in the game. Does not support multiple cameras (Just using what photonvision gives us) and uses WpiLib's AprilTags field layout

### CODE REVIEW

File: lib/build.gradle [hunk 1]
```diff
@@ -34,14 +29,15 @@ dependencies {
...
-    implementation 'edu.wpi.first.hal:hal-java:2022.+'
+    implementation 'edu.wpi.first.cscore:cscore-java:2023.2.1'
+    implementation 'edu.wpi.first.cameraserver:cameraserver-java:2023.2.1'
+    implementation 'edu.wpi.first.ntcore:ntcore-java:2023.2.1'
+    implementation 'edu.wpi.first.wpilibj:wpilibj-java:2023.2.1'
+    implementation 'edu.wpi.first.wpiutil:wpiutil-java:2023.2.1'
+    implementation 'edu.wpi.first.wpimath:wpimath-java:2023.2.1'
+    implementation 'edu.wpi.first.wpilibNewCommands:wpilibNewCommands-java:2023.2.1'
+    implementation 'edu.wpi.first.hal:hal-java:2023.2.1'
+    implementation "edu.wpi.first.apriltag:apriltag-java:2023.2.1"
```
Comment (tywilly) [inline lib/build.gradle:40]: Add a plus to the version to take the latest versions. Refer to the old versions for help

File: lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/AprilTags.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        List<PhotonTrackedTarget> targets = new ArrayList<>();
+
+        for(PhotonCamera camera : cameras) {
+            for(PhotonTrackedTarget target : camera.getLatestResult().getTargets()) {
+                boolean toAdd = true;
+                // check if the target is already in the list and if it has a higher pose ambiguity
+                for(PhotonTrackedTarget arTarget : targets) {
+                    if(target.getFiducialId() == arTarget.getFiducialId() && target.getPoseAmbiguity() < arTarget.getPoseAmbiguity()) {
+                        // if it does, remove it from the list
+                        targets.remove(arTarget);
```
Comment (tywilly) [inline lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/AprilTags.java:32]: Concurrent array modification. Create a temp array for operations

File: lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/AprilTags.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                    } else if(target.getFiducialId() == arTarget.getFiducialId()) {
+                        // if the target is already in the list and has the same ID, don't add it
+                        toAdd = false;
+                        break;
+                    }
+                }
+
+                if(toAdd) targets.add(target);
+            }
+            targets.addAll(camera.getLatestResult().getTargets());
```
Comment (tywilly) [inline lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/AprilTags.java:43]: ?????

File: lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/AprilTags.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        
+        return false;
+    }      
+
+    // method to get the closest target
+    public PhotonTrackedTarget getClosestTarget() {
+        if(!hasTargets()) return null;
+        PhotonTrackedTarget closestTarget = null;
+        for(PhotonTrackedTarget target : getTargets()) {
+            if(closestTarget == null || calculateDistance(target.getBestCameraToTarget()) > calculateDistance(target.getBestCameraToTarget()) ) closestTarget = target;
```
Comment (tywilly) [inline lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/AprilTags.java:64]: Checking difference between same targets

File: lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/AprilTags.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    // method to check if any camera has any targets
+    public boolean hasTargets() {
+        for(PhotonCamera camera : cameras) {
+            // if a camera has targets, return true
+            if(camera.getLatestResult().hasTargets()) return true;
+        } 
+        
+        return false;
```
Comment (tywilly) [inline lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/AprilTags.java:56]:
```suggestion
         // if a camera has targets, return true
         return cameras.stream().anyMatch((camera) -> camera.getLatestResult().hasTargets() == true);
```

File: lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/AprilTags.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        PhotonTrackedTarget lowestAmbiguity = null;
+        for(PhotonTrackedTarget target : getTargets()) {
+            // check if the target has a lower pose ambiguity than the current lowest ambiguity target
+            if(lowestAmbiguity == null || lowestAmbiguity.getPoseAmbiguity() > target.getPoseAmbiguity()) lowestAmbiguity = target;
+        }
+        return lowestAmbiguity;
+    }
+
+    private double calculateDistance(Transform3d t) {
+        return Math.sqrt(t.getX()*t.getX() + t.getY()*t.getY() + t.getZ()*t.getZ());
```
Comment (tywilly) [inline lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/AprilTags.java:81]: Remove Z dimension

And

This only calculates the distance from the passed in object and itself.

Camera is at 0,0 so try this

```suggestion
        return Math.sqrt(Math.pow(t.getX()-0,2) + Math.pow(t.getY()-0, 2));
```

Or

```suggestion
        return Math.sqrt(t.getX()*t.getX() + t.getY()*t.getY());
```

File: lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/RobotFieldPosition.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    /*
+        Example of Transform3d: 
+
+        Transform3d robotToCam = new Transform3d(new Translation3d(0.5, 0.0, 0.5), new Rotation3d(0,0,0)); 
+        Cam mounted facing forward, half a meter forward of center, half a meter up from center.
+    */
+    public RobotFieldPosition(String camera, Transform3d robotToCam) {
+        try {
+            layout = AprilTagFieldLayout.loadFromResource(AprilTagFields.k2022RapidReact.m_resourceFile);
```
Comment (tywilly) [inline lib/src/main/java/com/fairportrobotics/frc/poe/CameraTracking/RobotFieldPosition.java:28]: 2022??

**Review by tywilly — CHANGES_REQUESTED** [review]
Check posted comments for suggestions

**Comment by Declipsonator** [comment]
# Changes
### Freeze and Unfreeze Method
When using AprilTags across multiple lines like this:
```
    if(aprilTags.hasTargets()) {
      SmartDashboard.putNumber("Closest AprilTag ID", aprilTags.getClosestTarget().getFiducialId());
      SmartDashboard.putString("Closest AprilTag Position", aprilTags.getClosestTarget().getBestCameraToTarget().toString());

      SmartDashboard.putNumber("Lowest AprilTag Ambiguity ID", aprilTags.getLowestAmbiguityTarget().getFiducialId());
      SmartDashboard.putNumber("Lowest AprilTag Ambiguity Ambiguity", aprilTags.getLowestAmbiguityTarget().getPoseAmbiguity());

    }
```

The targets inside AprilTags can change since they are not synchronized, causing crashes. By using the methods like this:
```
    aprilTags.freeze();
    
    if(aprilTags.hasTargets()) {
      SmartDashboard.putNumber("Closest AprilTag ID", aprilTags.getClosestTarget().getFiducialId());
      SmartDashboard.putString("Closest AprilTag Position", aprilTags.getClosestTarget().getBestCameraToTarget().toString());

      SmartDashboard.putNumber("Lowest AprilTag Ambiguity ID", aprilTags.getLowestAmbiguityTarget().getFiducialId());
      SmartDashboard.putNumber("Lowest AprilTag Ambiguity Ambiguity", aprilTags.getLowestAmbiguityTarget().getPoseAmbiguity());

    }

    aprilTags.unFreeze();
```
All problems can be mitigated.

**Review by tywilly — APPROVED** [review]

**Pull request merged**